### DISCUSSION TITLE
Difficulty training top-down model with 3 mice

### DISCUSSION THREAD

**Discussion by timgjohnson** [talmolab/sleap#1369] [Help!]
Hi, we are relatively new to using SLEAP/ML tools and seem to have hit a performance ceiling. Trying to understand the extent to which we can improve performance by changing the training parameters. 

We have labeled 620 of ~40,000 frames in our video. We have been using the following skeleton and training parameters:
![skeleton](https://github.com/talmolab/sleap/assets/137902068/5a69201c-f1ad-485b-8be7-ffe9feb0b54c)
![image](https://github.com/talmolab/sleap/assets/137902068/c0e42f08-fbc7-495a-ab71-afe9d06a707b)
![image](https://github.com/talmolab/sleap/assets/137902068/4986992b-d923-48a9-91c9-23f845a285c0)
![image](https://github.com/talmolab/sleap/assets/137902068/2f74afb2-1e1b-462d-8c21-77c54db7b8ca)

Our resulting metrics typically look like the following for centroid and centered instance respectively:
![image](https://github.com/talmolab/sleap/assets/137902068/48584c02-09a7-4612-ba54-c9a66bb21d88)
![image](https://github.com/talmolab/sleap/assets/137902068/381d3c30-ac79-488b-84cb-49c7fd652ad3)

From running inference, we have observed that it mainly struggles with placing nodes across multiple animals when they are close to one another and when the mice are up against the wall. Are there parameter changes that would likely help with this issue?

We have also had some issues with memory allocation. We typically get a lot of warnings at the starts of runs saying that the GPU ran out of memory trying to allocate around 2-3 GB with freed_by_count=0. There is also a warning saying to check callbacks due to a discrepancy between the batch time and the on_train_batch_end time. The former set of warnings are highlighted in the image, and the latter is unhighlighted, and occupies the bottom two lines. 
![image](https://github.com/talmolab/sleap/assets/137902068/9a15015a-510f-4096-8125-f761e155ab5c)

Additionally, when we have tried training a centered instance model with a pretrained model, the program seems to crash (will get stuck and become unresponsive seemingly indefinitely). 

To what extent are these addressable by decreasing batch size, and is it worth the decrease in model performance? Alternatively would it be better to just run on colab to circumvent memory constraints? We tried it once but the average time per epoch was significantly slower, and it felt more difficult to know when to stop training without the GUI available.

We were also wondering whether it made more sense to curate our added frames more. So far, our general workflow has been run training, run inference on 20 frames, correct node placements on the inference frames, add them, and run again. Sometimes when the mice are climbing, they stick their nose over the edge or partially exit frames. Would there be any merit in omitting frames where mice are in those kinds of positions (especially as we just made a box to prevent them from doing that in subsequent videos)? Wondering if that would help improve the model's consistency.

Would appreciate any available insight.

**Comment by talmo** [comment]
Hi @timgjohnson,

Thanks for the comprehensive report on your setup! Super helpful for diagnostics. Let's get into it one by one:

> We have labeled 620 of ~40,000 frames in our video. 

Do you only have a single video? Is this just a test project? I'm asking because in general you'll want to maximize diversity in your dataset, which means labeling frames across more videos from your entire dataset that span dates, time of day, animals and experimental conditions. If >20-50% of your labels come from a single video, you'll start to get models that won't generalize to new videos very well.

> We have been using the following skeleton and training parameters:

Here's some suggestions:

_Centroid:_
- Set the rotation augmentation to +-180 since you have an overhead point of view

_Centered instance:_
- Enable rotation augmentation (already set to +-180 it looks like)
- Increase the max stride to 32. This will increase the receptive field size such that the model is better able to reason about larger spatial scales (see the blue box on the left of the model configuration window)
- Decrease the filters rate to 1.5. This balances having fewer larger scale features in favor of more small scale ones, but more importantly it decreases the overall number of parameters which are leading to your memory issues.
- Decrease the filters to 32 or 48. Same as above, this will decrease the number of parameters of the model, which right now are a bit much.
- Decrease the sigma to 2.5. While 5.0 makes it easier to learn coarser features, it'll result in issues when the animals are too close since the confidence maps will begin to overlap.

More generally, I'd also recommend giving the bottom-up approach a go as well since we've found that it can sometimes work better in mouse datasets.

[This troubleshooting guide](https://sleap.ai/guides/troubleshooting-workflows.html) may also be helpful for future reference.


> We have also had some issues with memory allocation. We typically get a lot of warnings at the starts of runs saying that the GPU ran out of memory trying to allocate around 2-3 GB with freed_by_count=0. There is also a warning saying to check callbacks due to a discrepancy between the batch time and the on_train_batch_end time. The former set of warnings are highlighted in the image, and the latter is unhighlighted, and occupies the bottom two lines.

Yep, this is because the model is too big to fit on your GPU, so TensorFlow is doing its best to try and optimize the memory layout to compensate. Try the suggestions above (especially `filters = 32` and `filters_rate = 1.5`) and see if you still get these.

> Additionally, when we have tried training a centered instance model with a pretrained model, the program seems to crash (will get stuck and become unresponsive seemingly indefinitely).

Same as above, except these models have an even bigger footprint that we can't really control (they're pretrained, so we can't reconfigure the network). You can try tweaking the upsampling stack parameters, but I'd be surprised if that works if you can't even fit the UNet on your GPU.

> To what extent are these addressable by decreasing batch size, and is it worth the decrease in model performance?

You can try decreasing the batch size, but the default of 4 is already quite minimal and lowering it might result in unstable training or weird model performance. Give it a go though -- sounds like an interesting experiment!

> Alternatively would it be better to just run on colab to circumvent memory constraints? We tried it once but the average time per epoch was significantly slower, and it felt more difficult to know when to stop training without the GUI available.

Yeah, Colab is not ideal and we have a plan to help address this via other cloud-based resources but this is at least a few months out.

In the meantime, if you have the budget I'd strongly recommend getting a new GPU with >6 GB of memory (all current gen NVIDIA GPUs like GeForce 3060 to 3080 and 40xx should meet this requirement). A great fully-integrated solution are the new [M1/M2 Mac Minis](https://www.amazon.com/Apple-Desktop-Computer-10%E2%80%91core-Ethernet/dp/B0BSHGHGXR/) which are surprisingly reasonably priced at <$600 and come with a strong hardware accelerator for deep learning which we now support natively.

> We were also wondering whether it made more sense to curate our added frames more. So far, our general workflow has been run training, run inference on 20 frames, correct node placements on the inference frames, add them, and run again. Sometimes when the mice are climbing, they stick their nose over the edge or partially exit frames. Would there be any merit in omitting frames where mice are in those kinds of positions (especially as we just made a box to prevent them from doing that in subsequent videos)? Wondering if that would help improve the model's consistency.

It makes a lot of sense to curate your added frames! The more diversity, the better, so definitely keep those frames where they're in weird positions. If it won't be an issue in future data, it might be best to treat this as a practice run and start fresh on those new videos, but either way it should work.


Let us know if you have any questions!

Cheers,

Talmo

**Reply by timgjohnson** [reply]
Hi Talmo,

Thank you for your in-depth reply! We've tried out many of your ideas and seen some improvement, but we have a few additional questions.

First, when we trained a top-down model with the parameters you suggested, the performance improved but we got a weird error distribution for our center node. It looks like the node placements are either dead-on, or off by specific distances. We are wondering if it could relate to our node placements (or something else on the human end) due to the apparent specificity of the error thresholds. 

![image](https://github.com/talmolab/sleap/assets/137902068/0bcd26ad-0482-4fce-9817-c3d688b3e56e)


We also tried training bottom-up, it performed similarly to the top-down model in terms of metrics. Do you have any recommended configs for bottom-up training? We just ran with defaults (besides adding +-180 for rotation) and are looking at optimizing the hyperparameters further.

![image](https://github.com/talmolab/sleap/assets/137902068/694cb976-36f5-40e6-a1e2-3be31294ba10)


We also had some questions about frame selection. We also noticed that 100+ labels are considered late-stage in the troubleshooting guide. Additionally, we noted that in the datasets on the website, the average frames per video seem much lower than what we have. How evenly were those labels distributed across those videos? Is it more optimal to label fewer, more specific frames over a larger number of videos? 

Thanks again!

**Reply by talmo** [reply]
Hi @timgjohnson,

Very, very strange that the center node is so off for top-down models! Also weird that it's at those discrete intervals. Anchor node predictions are usually pretty dead-on since you needed the centroid model to detect it in the first place. Are you definitely using the same anchor node ("Center") for both centroid and centered instance models?

Do you have any screenshots of a frame where the centroid/Center node prediction was off?

It is interesting that it's similarly an outlier in the bottom-up model. This model type should have no special association with the center node (unlike the top-down model which uses it as an anchor).

Given these observations, it'd really be worthwhile to go back and take a look at your manual labels to see if you can find anything that might be weird or inconsistent (like marking them as not visible or putting them in very inconsistent locations on the morphology of the animal).

> We also had some questions about frame selection. We also noticed that 100+ labels are considered late-stage in the troubleshooting guide.

Admittedly, 100+ is not necessarily late stage depending on the dataset. 1000+ is more realistic for complex datasets with multiple animals and enrichment or other objects you might find in the home cage.

> Additionally, we noted that in the datasets on the website, the average frames per video seem much lower than what we have. How evenly were those labels distributed across those videos? Is it more optimal to label fewer, more specific frames over a larger number of videos?

It's definitely the case that it'll be optimal to label fewer, more specific frames over a larger number of videos as this will maximize diversity in the training dataset. Similarly, it's best to avoid labeling frames that are temporally adjacent since they'll also be super similar to one another.


Let me know about the above suggestions and we'll take it from there.

Cheers,

Talmo

**Reply by timgjohnson** [reply]
Hi Talmo, 

Really appreciate all the help. Adding multiple videos has helped a bunch. Our maP scores increased by ~15% with just 50 frames from a second video! We're currently working on adding more footage.

![image](https://github.com/talmolab/sleap/assets/137902068/329650dc-46fe-42c1-87e6-2e3503857916)

For the weird center node error distribution, we were looking through the frames on the first video and it seems like we had different standards for labeling center at different points in time. We used to place them more posterior, and never went back and relabeled them. In the most recent run, it seems to have picked where it wants to place them, but based on the visualization it's more posterior than we would like.

We were wondering if it made sense to cull frames when going back and doing said relabeling? If the ideal setup is a collection of curated frames from a larger number of videos, then it seems like our current setup would weigh data from the first video more heavily due to the sheer number of labels. Additionally, a lot of those labels are on similar (redundant?) kinds of frames. Would love to understand whether culling/normalizing frame count makes sense.

Additionally, there are some differences in resolution and glare between the videos (images below). We are looking to play with the noise related augmentations to try and adjust for those differences. Are there different use cases for gaussian, uniform, or both. Alternatively, if these aren't likely to be helpful that would also be helpful to know. Putting a representative image from each video below.

![image](https://github.com/talmolab/sleap/assets/137902068/266bd4d0-4cce-423d-a3a7-8c92cbdc2885)
![image](https://github.com/talmolab/sleap/assets/137902068/6f379e1b-42a7-43a9-8699-1ae97ec4a884)
![image](https://github.com/talmolab/sleap/assets/137902068/ae6b6976-807d-4303-9b23-88dd9cacec59)

Thanks again for all the help!

**Reply by talmo** [reply]
> Hi Talmo,
> 
> Really appreciate all the help. Adding multiple videos has helped a bunch. Our maP scores increased by ~15% with just 50 frames from a second video! We're currently working on adding more footage.

Awesome!


> For the weird center node error distribution, we were looking through the frames on the first video and it seems like we had different standards for labeling center at different points in time. We used to place them more posterior, and never went back and relabeled them. In the most recent run, it seems to have picked where it wants to place them, but based on the visualization it's more posterior than we would like.

Yep, labeling conventions matter a lot. What we like to do is to create a slideshow or doc that has illustrated examples of the annotation protocol so we're both explicit and self-consistent. This also helps to promote reproducibility once you publish the work.


> We were wondering if it made sense to cull frames when going back and doing said relabeling? If the ideal setup is a collection of curated frames from a larger number of videos, then it seems like our current setup would weigh data from the first video more heavily due to the sheer number of labels. Additionally, a lot of those labels are on similar (redundant?) kinds of frames. Would love to understand whether culling/normalizing frame count makes sense.

If you have enough frames from other videos it might be ok to keep them in, but it's really an empirical question. In general, you'll likely get the best results by having balanced representation with less redundancy. That said, 3-5 images that look similar is helpful to capture subtle variations that you want the model to be agnostic to. 300-500 is probably overkill and will create an imbalanced dataset.


> Additionally, there are some differences in resolution and glare between the videos (images below). We are looking to play with the noise related augmentations to try and adjust for those differences. Are there different use cases for gaussian, uniform, or both. Alternatively, if these aren't likely to be helpful that would also be helpful to know. Putting a representative image from each video below.

Oo, not ideal but if you don't have control over that (e.g., the data already exists), then you can get around it with more labeling. Augmentation can help, but in our experience they don't end up being a sufficient substitute for more labels on the actual data since it's hard to faithfully model the statistical differences that result from different imaging conditions.

More generally regarding imaging conditions: if possible, I'd strongly advise lowering the exposure time on the camera. You get more photons and therefore higher contrast images when you have a longer exposure, but you also get more motion blur as you can see in third image. Motion blur makes pose estimation really challenging since it's hard to localize landmarks when the animal is smeared (even for a human). You'll usually get much better results by lowering the exposure time and increasing the illumination (by adding more lights/LEDs). You don't need to change the framerate if your camera supports changing the exposure independently. And definitely don't leave it on auto since that'll just introduce more variability!


Let us know if you have any other questions!

Cheers,

Talmo

**Reply by timgjohnson** [reply]
Hi Talmo,

Thank you again for your helpful replies! We've been adding videos, frames, and your suggestions and seen good results, but we seem to have hit a performance ceiling. We've seen very minimal improvement after adding 600 frames equally across ten videos (at 60 FPS). Our latest training settings and evaluation metrics are below:

![image](https://github.com/talmolab/sleap/assets/137902068/995ac972-6cef-460d-8f6a-edae0e56accb)
![image](https://github.com/talmolab/sleap/assets/137902068/8a2f514a-2481-4781-a3e6-6ed5654bc167)
![image](https://github.com/talmolab/sleap/assets/137902068/5bfe53b6-5757-4af4-80ad-97d72e1bd58a)
![image](https://github.com/talmolab/sleap/assets/137902068/87651601-4098-4b07-8793-82f0c3c34744)

We have also started using a box on the most recent four videos to help stabilize imaging conditions. Could this be a source of error since the boxed footage is inconsistent with our previous footage? We've considered removing the older footage and adding more boxed footage.
![image](https://github.com/talmolab/sleap/assets/137902068/64f241b9-9a6a-4ed7-8c39-8cdddcc4c7f4)

We've hit an OKS mAP ceiling at around 0.6. Do you have any recommendations for overcoming this barrier?

**Reply by talmo** [reply]
Hi @timgjohnson,

Nice! I still see that bimodal distribution of the `Center` node which is concerning due to the aforementioned labeling inconsistency strategy.

Great idea re: adding the box to stabilize imaging conditions.

You could try excluding the old footage and it'll probably do better on the newer more standardized videos, but if it's not hurting performance and they're similar enough I wouldn't worry about it.

I also wouldn't read too much into the mAP scores -- they're a rough guideline and sensitive to how many nodes you have. If you're happy with the poses 90-95%+ of the time, I'd move on to the tracking since that'll likely be the bigger challenge with 3 identical black mice 😅

While you're still in the imaging setup tweaking stage, I'd again recommend cranking that exposure down to reduce motion blur like this:
![image](https://github.com/talmolab/sleap/assets/3187454/d8a0bd25-69ce-4a64-8b9b-2b8ce3ba0d9d)

An exposure time of ~5ms is pretty good for reducing these if you can get away with it.

Let us know if you have any questions!

Cheers,

Talmo
